### PR TITLE
#171380782 Get most traveled destinations

### PR DESCRIPTION
**Pull request by descholar-ceo**
#### What does this PR ​ do​ ? 
Creates an endpoint for getting the most traveled destinations

#### Description of Task to be completed?

Have the following endpoint working 

`GET    /api/trips/most-traveled-destinations`

##### A.Pre-requisites : 
* To be able to test this PR locally, you need to have [PostgreSQL version `9.5.21` or higher](https://www.postgresql.org/) database engine already installed on your local machine or running in a container.
* You need to have [NodeJS version `6` or higher](https://nodejs.org/en/) and [npm version `6` or higher](https://www.npmjs.com/)
* You need to have [postman](https://www.postman.com/downloads/) installed on your local machine to be able to test the endpoint
* Your local machine needs to be able to run terminal, powershell, or bash commands

##### B.Steps:
1. If you don't yet have this repository on your local machine, clone it with this command in your terminal `git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend`
2. After cloning the repository, run the following command respectively
    > * cd stackup2-barefoot-backend
    > * git fetch origin ft-most-traveled-destinations-171380782
    > * git checkout ft-most-traveled-destinations-171380782
    > * npm install
3. Rename the file `.env.example` to `.env` which is in root directory of this project (current directory if you correctly followed all of the above steps respectively)
4. Fill the values of PORT, and DEVELOPMENT DATABASE VALUES
5. Then run the following commands in your terminal respectivelly:
    > * npm run create-db
    > * npm run migrations
    > * npm start
6. Signup, verify your account and Login, and paste the token you get in headers as authorization `Bearer token`
7. After all of the previous steps open your postman and send a GET request to `http://<your_server_hostname>:<server_port>/api/trips/most-traveled-destinations`
===>You should get this response 
```javascript
{
    "error": "There is no trip has done yet"
}
```
===>This is because there is no trip yet in database

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[ #171380782](https://www.pivotaltracker.com/n/projects/2435482/stories/171380782)
#### When there are some travels
Screenshot from 2020-03-22 21-56-51

##### When there is no travel yet
Screenshot from 2020-03-22 21-00-49

#### Questions:
N/A

### CODE REVIEW

File: src/controllers/request.controller.js [hunk 1]
```diff
@@ -31,4 +31,32 @@ export default class RequestController {
...
+      errorResponse(res, notFound, noTripYet);
+    }
+  }
+
+  /**
+   * @param {object} req
+   * @param {object} res
+   * @returns {object} sends response to user
+   */
+  static acceptRequest = async (req, res) => {
```
Comment (the22mastermind) [inline src/controllers/request.controller.js]: What does this method do?

File: src/database/migrations/20200312122638-create-request.js [hunk 2]
```diff
@@ -19,7 +19,7 @@ module.exports = {
       allowNull: false,
       type: Sequelize.STRING
     },
-    travelTo: {
+    destination: {
```
Comment (the22mastermind) [inline src/database/migrations/20200312122638-create-request.js]: Why did you change this attribute? 
Could you keep it as `travelTo` and when you are using it in a query you put it in quotes for eg. `SELECT "travelTo" FROM ...`?

File: src/database/models/request.js [hunk 3]
```diff
@@ -3,7 +3,7 @@ module.exports = (sequelize, DataTypes) => {
   const request = sequelize.define('request', {
     userId: DataTypes.INTEGER,
     travelFrom: DataTypes.STRING,
-    travelTo: DataTypes.STRING,
+    destination: DataTypes.STRING,
```
Comment (the22mastermind) [inline src/database/models/request.js]: Same here, keep it as `travelTo`

File: src/routes/api/request.routes.js [hunk 4]
```diff
@@ -12,5 +12,7 @@ const {
 const router = express.Router();
 
 router.post('/', [isCurrentUserLoggedIn, isCurrentUserVerified, validateTripRequest], createTripRequest);
+router.get('/most-traveled-destinations', isCurrentUserLoggedIn, isCurrentUserLoggedIn, placesAndVisitTimes);
```
Comment (the22mastermind) [inline src/routes/api/request.routes.js]: 1. Could you rename this route to maybe `/trips/statistics`?
2. Why have you put the same middleware `isCurrentUserLoggedIn` twice?

File: src/services/request.service.js [hunk 5]
```diff
@@ -12,12 +12,13 @@ export default class RequestService {
    * @returns {Object} Saved trip request details
     */
   static handleTripRequest(tripRequestData) {
+    console.log(tripRequestData);
```
Comment (the22mastermind) [inline src/services/request.service.js]: Maybe you forgot to remove this `console`?

File: src/services/request.service.js [hunk 6]
```diff
@@ -12,12 +12,13 @@ export default class RequestService {
...
     return request.create(
       tripRequestData,
       {
         fields: [
           'userId',
-          'travelTo',
+          'destination',
```
Comment (the22mastermind) [inline src/services/request.service.js]: Check comment above in migrations/models

File: src/services/request.service.js [hunk 7]
```diff
@@ -27,4 +28,27 @@ export default class RequestService {
...
     );
   }
+
+  /**
+   * @returns {Integer} returns number of traveled places
+   * @description This function getMostTraveledDestinations()
+   *  returns all of the destinations a and their visit times
+   */
+  static getMostTraveledDestinations = async () => {
+    const placeAndTheirVisitTimes = await sequelize.query('SELECT destination, COUNT(destination) FROM requests WHERE status !=\'pending\' AND status!=\'rejected\' GROUP BY destination ORDER BY count DESC');
```
Comment (the22mastermind) [inline src/services/request.service.js]: You will need to modify the column you're selecting

File: src/services/request.service.js [hunk 8]
```diff
@@ -27,4 +28,27 @@ export default class RequestService {
...
+    
+    return placeAndTheirVisitTimes;
+  }
+
+
+   /**
+   * @param {Integer} id number of traveled places
+   * @returns {Integer} returns number of traveled places
+   * @description This function getMostTraveledDestinations()
```
Comment (the22mastermind) [inline src/services/request.service.js]: Is this description accurate?

File: src/utils/customMessages.js [hunk 9]
```diff
@@ -51,7 +51,9 @@ const messages = {
   usernameExistOrEmpty: 'The username you entered already exists or is empty, if you want to continue update your profile, change the new username or unselect it to remain with the old one',
   passwordChangeSuccess: 'Your password has been changed, login with new password',
   changeEmailNotAllowed: 'You cannot change your email, unselect it if you want to continue updating your profile, otherwise you will not be able to continue',
-  tokenInvalid: 'Authentication failed. Token is invalid'
+  tokenInvalid: 'Authentication failed. Token is invalid',
+  placesRetrieved: 'Successfully retrieved destinations and their visit times',
+  noTripYet: 'There is no trip has done yet'
```
Comment (the22mastermind) [inline src/utils/customMessages.js]: Maybe you should change this to `noPlacesRetrieved: 'Data not available for now, please try again later'`

File: src/utils/validators.js [hunk 10]
```diff
@@ -37,7 +37,7 @@ static async validateOneWayTripRequest(tripRequestData) {
   const schema = Joi.object({
       travelFrom: cleanString
       .messages(createValidationErrors('string', invalidTravelFrom)),
-      travelTo: cleanString
+      destination: cleanString
```
Comment (the22mastermind) [inline src/utils/validators.js]: Same comment as in models

**Review by the22mastermind — COMMENTED** [review]
@descholar-ceo Good job. 
I left you some comments and suggestions.
Keep it up.

File: src/controllers/request.controller.js (hunk 1 above)
Comment (descholar-ceo) [reply src/controllers/request.controller.js]: @the22mastermind  This method is to help me approve a request to be able to pass tests, actually, approving request is not my task, but I need it because the default request status is `pending` so because of that I missed a way of testing if the route that I am working is working, this method will be deleted or modified with someone who will do the request approval, but now is needed for testing purpose!

File: src/database/migrations/20200312122638-create-request.js (hunk 2 above)
Comment (descholar-ceo) [reply src/database/migrations/20200312122638-create-request.js]: @the22mastermind Thank you so much for this suggestion of yours, I implemented it and it is working, without changing other people's work, thank you so much, I didn't know that I'd use `"travelTo"`, because using `travelTo` was throwing an error, but now is okay! Thank you a lot my brother

File: src/database/models/request.js (hunk 3 above)
Comment (descholar-ceo) [reply src/database/models/request.js]: I changed It

File: src/routes/api/request.routes.js (hunk 4 above)
Comment (descholar-ceo) [reply src/routes/api/request.routes.js]: 1. If I change this route to `/trips/statistics` , what about this task ? 
Screenshot from 2020-03-23 09-40-54
I think it the one with statistics, so it would bring some confusion, maybe if you have any other suggestion apart from statistics I would follow that.

2. Adding `isCurrentUserLoggedIn` twice, was a mistake, I change it

File: src/services/request.service.js (hunk 5 above)
Comment (descholar-ceo) [reply src/services/request.service.js]: Yes for sure, I had removed it

File: src/services/request.service.js (hunk 6 above)
Comment (descholar-ceo) [reply src/services/request.service.js]: I change this one

File: src/services/request.service.js (hunk 7 above)
Comment (descholar-ceo) [reply src/services/request.service.js]: I change it too

File: src/services/request.service.js (hunk 8 above)
Comment (descholar-ceo) [reply src/services/request.service.js]: I changed it, made it short but descriptive enough

File: src/utils/validators.js (hunk 10 above)
Comment (descholar-ceo) [reply src/utils/validators.js]: I change it

File: src/utils/customMessages.js (hunk 9 above)
Comment (descholar-ceo) [reply src/utils/customMessages.js]: I changed it and made it even better for description

**Comment by descholar-ceo** [comment]
> @descholar-ceo Good job.
> I left you some comments and suggestions.
> Keep it up.

@the22mastermind I have implemented all of the feedback you gave me.

**Review by Muhire-Josue — APPROVED** [review]
LGTM

File: src/routes/api/request.routes.js [hunk 11]
```diff
@@ -12,5 +12,7 @@ const {
 const router = express.Router();
 
 router.post('/', [isCurrentUserLoggedIn, isCurrentUserVerified, validateTripRequest], createTripRequest);
+router.get('/most-traveled-destinations', isCurrentUserLoggedIn, isCurrentUserVerified, placesAndVisitTimes);
+router.get('/accept-request/:id', acceptRequest);
```
Comment (since9629) [inline src/routes/api/request.routes.js]: @descholar-ceo since you're updating the database, could you use 'PATCH' instead of 'GET' and use request body to hold the `id` field

File: src/services/request.service.js (hunk 8 above)
Comment (since9629) [reply src/services/request.service.js]: I still can't see the change in its description

File: src/routes/api/request.routes.js (hunk 11 above)
Comment (descholar-ceo) [reply src/routes/api/request.routes.js]: @karamuka I implemented this feedback

**Review by since9629 — APPROVED** [review]

File: src/database/migrations/20200312122638-create-request.js (hunk 2 above)
Comment (the22mastermind) [reply src/database/migrations/20200312122638-create-request.js]: You're welcome bro.

File: src/routes/api/request.routes.js (hunk 4 above)
Comment (the22mastermind) [reply src/routes/api/request.routes.js]: Ok. You're right. It makes sense to leave it as `/most-traveled-destinations`

File: src/services/request.service.js (hunk 8 above)
Comment (descholar-ceo) [reply src/services/request.service.js]: @karamuka I changed it, if you go in codes, you can find it, I don't know why isn't appearing here on comments
Screenshot from 2020-03-23 15-48-07

**Review by the22mastermind — COMMENTED** [review]
@descholar-ceo Good job. But I've noticed that your `acceptRequest` method doesn't have any validation, could you at implement at least a validation for checking if that `reqId` is provided or not?
Other than that, it looks good.

**Comment by descholar-ceo** [comment]
> @descholar-ceo Good job. But I've noticed that your `acceptRequest` method doesn't have any validation, could you at implement at least a validation for checking if that `reqId` is provided or not?
> Other than that, it looks good.

@the22mastermind I implemented this too

**Review by Ugizwenayo-Divine — APPROVED** [review]
you did great job

**Review by the22mastermind — APPROVED** [review]
Well done @descholar-ceo 
Keep up the good work!

**Review by muchristian — APPROVED** [review]
well done Emmanuel

File: src/services/request.service.js [hunk 12]
```diff
@@ -27,4 +27,25 @@ export default class RequestService {
       }
     );
   }
+
+  /**
+   * @returns {Integer} returns number of traveled places
+   * @description returns all of the destinations a and their appearance times from the database
+   */
+  static getMostTraveledDestinations = async () => {
+    const placeAndTheirVisitTimes = await sequelize.query('SELECT "travelTo", COUNT(*) FROM requests WHERE status !=\'pending\' AND status!=\'rejected\' GROUP BY "travelTo" ORDER BY count DESC');
```
Comment (mifeille) [inline src/services/request.service.js]: Good job @de-scholar but I think that here you should use where status= 'approved' cause currently when I change the status to something which is not `pending` or `rejected` like `sdabdhbfbfd` in my database it lists that trip on that destinations list you are returning.

Comment (descholar-ceo) [reply src/services/request.service.js]: I changed this

**Review by mifeille — APPROVED** [review]
